### PR TITLE
added level progress and offers

### PR DESCRIPTION
**Pull request by mserranom**

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -3,7 +3,13 @@
 set -e
 
 echo "installing dependencies..."
-pipenv install --dev
+pipenv install --dev --ignore-pipfile
```
Comment (mserranom) [inline build.sh]: Adding `ignore-pipfile` installs dependencies from the lockfile. This means:
1. remote and local builds are identical
2. lockfile won't be updated unless we manually force so (by running specific pipenv commands)

File: event_data/generate_session.py [hunk 2]
```diff
@@ -1,44 +1,112 @@
...
                 device_atts):
 
-    session_duration_seconds = round(random.expovariate(1.5) * 100 + 5)
-    game_start_delta_seconds = random.randint(8, 30)
+    session_id = Cryptographic().uuid()
+
+    events = list()
+
+    # TODO: find a better way to keep time accumulator
+    current_time = {"time": start_time}
```
Comment (mserranom) [inline event_data/generate_session.py]: If I make this an integer, then the variable is not found under `gen_core_atts` function. (line 52). Not sure what's the best way to do this, any help?

Comment (vigbk) [reply event_data/generate_session.py]: https://docs.python.org/2/faq/programming.html#why-am-i-getting-an-unboundlocalerror-when-the-variable-has-a-value

Comment (mserranom) [reply event_data/generate_session.py]: ok, understood. So current solution looks good right @vigbk ? I'll remove the TODO if so.

File: event_data/generate_attributes.py [hunk 3]
```diff
@@ -22,3 +25,49 @@ def gen_device_attributes():
...
+    rand = random.random()
+    if rand > 0.95:
+        return "gold"
+    elif rand > 0.8:
+        return "silver"
+    else:
+        return "bronze"
+
+
+def gen_game_attributes(*,
```
Comment (prashastk) [inline event_data/generate_attributes.py]: How is the * as the first argument used here?

File: event_data/generate_attributes.py [hunk 4]
```diff
@@ -22,3 +25,49 @@ def gen_device_attributes():
...
+        "level": level,
+        "score": score,
+        "badge": badge,
+        "won": won
+    }
+
+
+merchandise_item_ids = range(4)
+merchandise_item_prices = [.99, 2.99, 5.99, 9.99]
+merchandise_item_categories = ["cat1", "cat2"]
```
Comment (prashastk) [inline event_data/generate_attributes.py]: Use all CAPS for constants?

File: event_data/generate_session.py [hunk 5]
```diff
@@ -1,44 +1,112 @@
...
+
+def popup_delay():
+    return 1 + round(random.random() * 3)
+
+
+def decision_delay():
+    return 5 + round(random.random() * 4)
+
 
 def gen_session(customer_identifier, start_time, user_atts, app_atts,
```
Comment (prashastk) [inline event_data/generate_session.py]: user_id is probably a lot more common way of specifying customer_identifier? (here and in the output as well).

File: event_data/generate_session.py (hunk 2 above)
Comment (vigbk) [reply event_data/generate_session.py]: yup

File: event_data/generate_attributes.py (hunk 3 above)
Comment (mserranom) [reply event_data/generate_attributes.py]: When the function accept more than 3 parameters adding `*` to enforce using parameter names makes code more clear.

It comes from the clean code rule of no more than 3 parameters.

Any thought?

File: event_data/generate_attributes.py (hunk 4 above)
Comment (mserranom) [reply event_data/generate_attributes.py]: will do

File: event_data/generate_session.py (hunk 5 above)
Comment (mserranom) [reply event_data/generate_session.py]: I took it from the spec, should I change it?

Comment (prashastk) [reply event_data/generate_session.py]: Up to you. My view is coloured by ads where customer (advertiser) and user (end-user) meant two different groups.

Comment (mserranom) [reply event_data/generate_session.py]: Just merged it. It needs a review anyway, so I'll keep it like this

File: event_data/generate_attributes.py (hunk 3 above)
Comment (prashastk) [reply event_data/generate_attributes.py]: Sounds good. I like the idea of doing this for more than 3 params!

**Pull request merged**